himalaya/internal/repair/repair.go

```diff
@@ -0,0 +1,365 @@
+)
+
+const (
+	limitMetadataLength = swarm.ChunkSize
+)
+
+// ProgressUpdater is and interface which can be implemented by client to recieve
+// updates from the utility
+type ProgressUpdater interface {
+	Update(string)
+}
+
+// Option is used to supply functional options for the repairer utility
+type Option func(*Repairer)
+
+// WithAPIStore is used to configure the API endpoint for running the utility. This
+// could be locally running bee node or some gateway
+func WithAPIStore(host string, port int, useSSL bool) Option {
+	return func(c *Repairer) {
+		c.store = cmdfile.NewApiStore(host, port, useSSL)
+	}
+}
+
+// WithLogger is used to supply optional logger to see debug messages
+func WithLogger(l logging.Logger) Option {
+	return func(c *Repairer) {
+		c.logger = l
+	}
+}
+
+// WithEncryption is used to enable encryption while creating data
+func WithEncryption(val bool) Option {
+	return func(c *Repairer) {
+		c.encrypt = val
+	}
+}
+
+// WithProgressUpdater is used to provide updater implementation to see updates
+// from utility
+func WithProgressUpdater(upd ProgressUpdater) Option {
+	return func(c *Repairer) {
+		c.updater = upd
+	}
+}
+
+// FileRepair takes in an older file reference and creates a new manifest which contains
+// the file and the metadata. This reference can be then used to query the /bzz endpoint to
+// serve the file
+//
+// Old Entry:
+// collection -> file reference -> file bytes
+//           |
+//           |-> metadata reference -> metadata bytes
+//
+// New Entry:
+// mantaray manifest -> Root Node (\) -> Metadata (index file)
+//                  |
+//                  |-> file entry -> Metadata (Filename, ContentType)
+//                                |
+//                                |-> File reference
+//
+func FileRepair(ctx context.Context, addr swarm.Address, opts ...Option) (swarm.Address, error) {
+	r := newWithOptions(opts...)
+
+	oldEntry, err := r.getOldFileEntry(ctx, addr)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	r.updater.Update(fmt.Sprintf("Updating reference for file %s", oldEntry.mtdt.Filename))
+
+	newManifest, err := manifest.NewDefaultManifest(r.ls, false)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	err = newManifest.Add(ctx, manifest.RootPath, manifest.NewEntry(
+		swarm.ZeroAddress,
+		map[string]string{
+			manifest.WebsiteIndexDocumentSuffixKey: oldEntry.mtdt.Filename,
+		},
+	))
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	err = newManifest.Add(
+		ctx,
+		oldEntry.mtdt.Filename,
+		manifest.NewEntry(oldEntry.e.Reference(), map[string]string{
+			manifest.EntryMetadataFilenameKey:    oldEntry.mtdt.Filename,
+			manifest.EntryMetadataContentTypeKey: oldEntry.mtdt.MimeType,
+		}),
+	)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	newReference, err := newManifest.Store(ctx)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	r.logger.Debugf("Created new file manifest with reference %s", newReference.String())
```

Comment:
maybe use `r.updater.Update`?

Reply:
The reason I did not add this final message in File, as well as Directory repair, is because the command prints out this statement at the end anyway.

At the end of the command in the command code we will print a statement saying:

[Created new reference <>] => This would be redundant I feel
Repaired file reference. New reference: <>